scrapy/pipelines/media.py

```diff
@@ -112,13 +147,19 @@ def _key_for_pipe(
     @classmethod
     def from_crawler(cls, crawler: Crawler) -> Self:
         pipe: Self
-        try:
+        if hasattr(cls, "from_settings"):
             pipe = cls.from_settings(crawler.settings)  # type: ignore[attr-defined]
-        except AttributeError:
+        elif "crawler" in get_func_args(cls.__init__):
+            pipe = cls(crawler=crawler)
+        else:
             pipe = cls()
-        pipe.crawler = crawler
-        assert crawler.request_fingerprinter
-        pipe._fingerprinter = crawler.request_fingerprinter
+            warnings.warn(
+                f"{cls.__qualname__}.__init__() doesn't take a crawler argument."
```

Comment:
In cases like these where the stacklevel would not help to point to the offending like, I wonder if we should prefix `cls.__qualname__` with the module name, i.e. use [`globa_object_name`](https://github.com/scrapy/scrapy/blob/261c4b61dc48353346c1e0387d0783ac15ab459d/scrapy/utils/python.py#L330).